lib/wasi.js

```diff
@@ -0,0 +1,89 @@
+'use strict';
+/* global WebAssembly */
+const { Array, ArrayPrototype, Object } = primordials;
+const { ERR_INVALID_ARG_TYPE } = require('internal/errors').codes;
+const { emitExperimentalWarning } = require('internal/util');
+const { WASI: _WASI } = internalBinding('wasi');
+const kSetMemory = Symbol('setMemory');
+
+emitExperimentalWarning('WASI');
+
+
+class WASI {
+  constructor(options = {}) {
+    if (options === null || typeof options !== 'object')
+      throw new ERR_INVALID_ARG_TYPE('options', 'object', options);
+
+    // eslint-disable-next-line prefer-const
+    let { args, env, preopens } = options;
+
+    if (Array.isArray(args))
+      args = ArrayPrototype.map(args, (arg) => { return String(arg); });
+    else if (args === undefined)
+      args = [];
+    else
+      throw new ERR_INVALID_ARG_TYPE('options.args', 'Array', args);
+
+    const envPairs = [];
+
+    if (env !== null && typeof env === 'object') {
+      for (const key in env) {
+        const value = env[key];
+        if (value !== undefined)
+          envPairs.push(`${key}=${value}`);
+      }
+    } else if (env !== undefined) {
+      throw new ERR_INVALID_ARG_TYPE('options.env', 'Object', env);
+    }
+
+    const preopenArray = [];
+
+    if (typeof preopens === 'object' && preopens !== null) {
+      Object.keys(preopens).forEach((key) => {
+        preopenArray.push(String(key));
+        preopenArray.push(String(preopens[key]));
+      });
+    } else if (preopens !== undefined) {
+      throw new ERR_INVALID_ARG_TYPE('options.preopens', 'Object', preopens);
+    }
+
+    const wrap = new _WASI(args, envPairs, preopenArray);
+
+    for (const prop in wrap) {
+      wrap[prop] = wrap[prop].bind(wrap);
```

Comment:
IIUC you could add a variant of `Environment::SetProtoMethod()` (and maybe `Environment::SetProtoMethodNoSideEffect()`) that’s called `Environment::SetInstanceMethod()` and does the exact same thing except with `InstanceTemplate()` instead of `PrototypeTemplate()`, and then use that instead of manually re-binding? Just a suggestion, but it seems like that might be a bit nicer?

Reply:
I added `Environment::SetInstanceMethod()` for `_setMemory()`, but the actual WASI APIs still need to be bound AFAICT because they are called with `this` set to `undefined` otherwise.